common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: RwLock<HashMap<Key, Arc<Mutex<Bucket>>>>,
+    default_bucket_size: usize,
+    default_fill_rate: usize,
+    open: bool,
+}
+
+#[allow(dead_code)]
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(default_bucket_size: usize, default_fill_rate: usize) -> Self {
+        // TODO: NonZeroUsize?
+        assert!(default_fill_rate > 0, "Fill rate must be greater than 0");
+        assert!(
+            default_bucket_size > 0,
+            "Bucket size must be greater than 0"
+        );
+        assert!(
+            default_bucket_size >= default_fill_rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open() -> Self {
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: std::usize::MAX,
+            default_fill_rate: std::usize::MAX,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&mut self, key: Key) -> Arc<Mutex<Bucket>> {
+        if !self.open {
+            self.bucket_inner(key, |size, rate| {
+                Arc::new(Mutex::new(Bucket::new(size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |_, _| Arc::new(Mutex::new(Bucket::open())))
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(usize, usize) -> Arc<Mutex<Bucket>>>(
+        &mut self,
+        key: Key,
+        bucket_create: F,
+    ) -> Arc<Mutex<Bucket>> {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        let maybe_value = { self.buckets.read().unwrap().get(&key).cloned() };
+
+        if let Some(value) = maybe_value {
+            value
+        } else {
+            let size = self.default_bucket_size;
+            let rate = self.default_fill_rate;
+            // Write in a bucket, but make sure again that it isn't there first
+            self.buckets
+                .write()
+                .unwrap()
+                .entry(key)
+                .or_insert_with(|| bucket_create(size, rate))
+                .clone()
+        }
+    }
+
+    /// Fill the buckets with the given rate, saturating at the max size
+    /// This should occur on an interval basis
+    fn refill(&mut self) {
+        self.buckets
+            .write()
+            .unwrap()
+            .values_mut()
+            .for_each(|bucket| {
+                bucket.lock().unwrap().refill();
+            });
+    }
```

Comment:
if `acquire_tokens` does a `refill`, is this necessary?

Reply:
This was originally to have periodic fills of the buckets at a regular interval.  But, now since it's doing it on acquire tokens it's not needed.